File: python/check_if_a_string_contains_all_binary_codes_of_size_k.py
class Solution:
    def hasAllCodes(self, s: str, k: int) -> bool:
        combSet: set[str] = set()
        for i in range(0, len(s) - k + 1):
            a = s[i : i + k]
            combSet.add(s[i : i + k])

        return len(combSet) == pow(2, k)


def test_hasAllCodes_case_1():
    # arrange
    s: str = "00110110"
    k: int = 2
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.hasAllCodes(s, k)

    # assert
    assert expected == actual


def test_hasAllCodes_case_2():
    # arrange
    s: str = "0110"
    k: int = 1
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.hasAllCodes(s, k)

    # assert
    assert expected == actual


def test_hasAllCodes_case_3():
    # arrange
    s: str = "0110"
    k: int = 2
    expected: bool = False

    # act
    solution = Solution()
    actual = solution.hasAllCodes(s, k)

    # assert
    assert expected == actual


def test_hasAllCodes_case_4():
    # arrange
    s: str = "00110"
    k: int = 2
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.hasAllCodes(s, k)

    # assert
    assert expected == actual

File: rust/src/code/check_if_a_string_contains_all_binary_codes_of_size_k.rs
struct Solution {}

impl Solution {
    pub fn has_all_codes(s: String, k: i32) -> bool {
        use std::collections::HashSet;

        let s: Vec<char> = s.chars().collect();
        let k: usize = k as usize;
        if s.len() + 1 < k {
            return false;
        }

        let mut comb_set: HashSet<String> = HashSet::new();

        for i in 0..(s.len() - k + 1) {
            comb_set.insert(s[i..(i + k)].iter().collect::<String>());
        }

        return comb_set.len() == 2_i32.pow(k as u32) as usize;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn has_all_codes_case_1() {
        // arrange
        let s: String = "00110110".to_string();
        let k: i32 = 2;
        let expected = true;

        // act
        let actual = Solution::has_all_codes(s, k);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn has_all_codes_case_2() {
        // arrange
        let s: String = "0".to_string();
        let k: i32 = 20;
        let expected = false;

        // act
        let actual = Solution::has_all_codes(s, k);

        // assert
        assert_eq!(expected, actual);
    }
}
